### ISSUE TITLE
Missing Values in CSV is taking as blank String not null

### ISSUE THREAD

**Issue by abhishek-ch** [databricks/spark-csv#86]
I am trying to load the csv file using pyspark and found blank values in CSV is always taken as blank String in Dataframe created even in PERMISSIVE mode but the doc says 

>PERMISSIVE: tries to parse all lines: nulls are inserted for missing tokens and extra tokens are ignored.

I am using following script -
```
df_test= sqlContext.read.format("com.databricks.spark.csv").options(delimiter=",").options(header="true").load("/Volumes/test.csv")
```
![image](https://cloud.githubusercontent.com/assets/7579608/8291106/3ce9a6e6-191f-11e5-8ff7-22e8d0fe4159.png)

thats why now fillna is not working .

**Comment by falaki** [comment]
What does the raw data look like? Just want to make sure there is no space or other character in place of that token.

**Comment by abhishek-ch** [comment]
its a blank cell with no space & character in a csv cell.

There is a piece of code in CSVReader
settings.setNullValue("")

I tried modifying it to null and it didn't work , it was bit hard to understand the code as not aware of scala.

**Comment by 1stdayonthejob** [comment]
I'm using the cars.csv from the readme and am also getting blank results.

~~~python
df = sqlContext.read.format('com.databricks.spark.csv').options(header='true').load('cars.csv')
df.fillna('SHOULD BE NULL').show()
~~~

~~~
+----+-----+-----+--------------------+--------------+
|year| make|model|             comment|         blank|
+----+-----+-----+--------------------+--------------+
|2012|Tesla|    S|          No comment|              |
|1997| Ford| E350|Go get one now th...|SHOULD BE NULL|
|2015|Chevy| Volt|      SHOULD BE NULL|SHOULD BE NULL|
+----+-----+-----+--------------------+--------------+
~~~

I would expect the 'blank' column in the Tesla line to be null.

Also if I add commas in the cars.csv file to have it separate all columns

~~~
year,make,model,comment,blank
"2012","Tesla","S","No comment",

1997,Ford,E350,"Go get one now they are going fast",
2015,Chevy,Volt,,
~~~

then all the fields are emty string rather than nulls

~~~
+----+-----+-----+--------------------+-----+
|year| make|model|             comment|blank|
+----+-----+-----+--------------------+-----+
|2012|Tesla|    S|          No comment|     |
|1997| Ford| E350|Go get one now th...|     |
|2015|Chevy| Volt|                    |     |
+----+-----+-----+--------------------+-----+
~~~

**Comment by JoshRosen** [comment]
I guess that we're only able to distinguish between empty strings and nulls if we assume that all empty strings will be quoted in the CSV?

**Comment by gregrahn** [comment]
+1 on this being a bug and some more thought needed in spark-csv on how to correctly handle  NULL vs empty string.  This simple example demonstrates the inconsistencies of this package.

Simply put, it should support reading a table from a source (like JDBC), writing that table out to CSV, read that data back, and have the exact same data that was in the original source.  This seems impossible today.

```
scala> val df = sqlContext.read.format("jdbc").options(Map(
     |              "url" -> "jdbc:postgresql://localhost:5432/grahn",
     |              "user" -> "grahn",
     |              "password" -> "",
     |              "dbtable" -> "public.spark",
     |              "driver" -> "org.postgresql.Driver")).load()
df: org.apache.spark.sql.DataFrame = [a: string, b: string]

scala> df.show()
+-----+----+
|    a|   b|
+-----+----+
|EMPTY|    |
| NULL|null|
+-----+----+

scala> df.registerTempTable("df")

scala> sqlContext.sql("SELECT * FROM df WHERE b is null").show()
+----+----+
|   a|   b|
+----+----+
|NULL|null|
+----+----+

scala> df.write.format("com.databricks.spark.csv").save("/tmp/spark-csv.csv")

scala> val df2 = sqlContext.read.format("com.databricks.spark.csv").load("/tmp/spark-csv.csv")
df2: org.apache.spark.sql.DataFrame = [C0: string, C1: string]

scala> df2.show()
+-----+----+
|   C0|  C1|
+-----+----+
|EMPTY|    |
| NULL|null|
+-----+----+

scala> df2.registerTempTable("df2")

scala> sqlContext.sql("SELECT * FROM df2 WHERE C1 is null").show()
+---+---+
| C0| C1|
+---+---+
+---+---+

scala> sqlContext.sql("SELECT * FROM df2 WHERE C1 = 'null'").show()
+----+----+
|  C0|  C1|
+----+----+
|NULL|null|
+----+----+

$ cat /tmp/spark-csv.csv/part-00000
EMPTY,
NULL,null
```

**Comment by honzasterba** [comment]
It woudl be awesome if loading nulls was possible, either by forcing empty strings to be quoted or being able to denote a special value (e.g. null or NA) to represent nulls in the csv

**Comment by rkkolla** [comment]
I am not able to apply fillna on my dataframe. We are using spark 1.3.1. Is there any why we can replace all null values to blank for all columns ?

**Comment by cherah30** [comment]
I spent more than 2 hours as you tested the thing and the only solution I have found is: 
my_dataframe.na.replace ('', 'unknown'). 

Note that I used spark 1.5.1 and syntax is that of PySpark.

**Comment by barrybecker4** [comment]
In 1.4.0 it looks like there is a "nullValue" option that allows you to specify (specificy?) the string that you want to use for null. This is great. It would be even greater if you could specify a list of such strings that should be interpreted as null. For example, 
option("nullValue", ["NA", "Missing", "MISSING", "", "?"])

**Comment by kadrach** [comment]
https://github.com/databricks/spark-csv/issues/192 seems to suggest this `nullValue` is only ever used when writing, not when reading.

**Comment by HyukjinKwon** [comment]
@kadrach No, `nullValue` is also supported when reading as described in `README.md`.

**Comment by HyukjinKwon** [comment]
I believe this is fixed in https://github.com/databricks/spark-csv/pull/224. Can we maybe close?

**Comment by barrybecker4** [comment]
nullValue is used when reading and writing. From my point of view, it is fine to close.

**Comment by gauravkhare** [comment]
df.na.replace(raw.columns,Map("" -> "0")).show()             // to convert from blank strings to zero

**Comment by Jasper-M** [comment]
For future generations who feel the urge to throw their computer out of the window:

```scala
.option("treatEmptyValuesAsNulls", "true")
```

**Comment by firefoxwang** [comment]
@HyukjinKwon 
'nullValue is also supported when reading'  is also doesn't work when I read csv file  with spark2.3, although I can use fillna instead it. Comments may need to be changed, the default nullValue is null and not an empty string
